File: packages/material-system/src/Material.ts
/**
 * Material class for managing rendering properties.
 * 用于管理渲染属性的材质类。
 */

import { BlendMode, CullMode, UniformType, UniformValue, MaterialDefinition } from './types';

/**
 * Material class that holds shader reference and uniform values.
 * 持有着色器引用和uniform值的材质类。
 */
export class Material {
    /** Material ID assigned by the engine | 引擎分配的材质ID */
    private _id: number = -1;

    /** Material name for debugging | 材质名称（用于调试） */
    public name: string;

    /** Shader ID to use | 使用的着色器ID */
    public shaderId: number;

    /** Blend mode | 混合模式 */
    public blendMode: BlendMode;

    /** Cull mode | 剔除模式 */
    public cullMode: CullMode;

    /** Depth test enabled | 是否启用深度测试 */
    public depthTest: boolean;

    /** Depth write enabled | 是否启用深度写入 */
    public depthWrite: boolean;

    /** Material uniform values | 材质uniform值 */
    private uniforms: Map<string, UniformValue> = new Map();

    /** Texture references by uniform name | 按uniform名称引用的纹理 */
    private textures: Map<string, string> = new Map();

    /** Whether the material needs to sync with engine | 材质是否需要与引擎同步 */
    private _dirty: boolean = true;

    constructor(name: string = 'Material', shaderId: number = 0) {
        this.name = name;
        this.shaderId = shaderId;
        this.blendMode = BlendMode.Alpha;
        this.cullMode = CullMode.None;
        this.depthTest = false;
        this.depthWrite = false;
    }

    /** Get the material ID | 获取材质ID */
    get id(): number {
        return this._id;
    }

    /** Set the material ID (called by MaterialManager) | 设置材质ID（由MaterialManager调用） */
    set id(value: number) {
        this._id = value;
    }

    /** Check if material is dirty | 检查材质是否为脏 */
    get dirty(): boolean {
        return this._dirty;
    }

    /** Mark material as clean | 标记材质为干净 */
    markClean(): void {
        this._dirty = false;
    }

    // ============= Uniform Setters =============
    // ============= Uniform设置方法 =============

    /**
     * Set a float uniform.
     * 设置浮点uniform。
     */
    setFloat(name: string, value: number): this {
        this.uniforms.set(name, { type: UniformType.Float, value });
        this._dirty = true;
        return this;
    }

    /**
     * Set a vec2 uniform.
     * 设置vec2 uniform。
     */
    setVec2(name: string, x: number, y: number): this {
        this.uniforms.set(name, { type: UniformType.Vec2, value: [x, y] });
        this._dirty = true;
        return this;
    }

    /**
     * Set a vec3 uniform.
     * 设置vec3 uniform。
     */
    setVec3(name: string, x: number, y: number, z: number): this {
        this.uniforms.set(name, { type: UniformType.Vec3, value: [x, y, z] });
        this._dirty = true;
        return this;
    }

    /**
     * Set a vec4 uniform.
     * 设置vec4 uniform。
     */
    setVec4(name: string, x: number, y: number, z: number, w: number): this {
        this.uniforms.set(name, { type: UniformType.Vec4, value: [x, y, z, w] });
        this._dirty = true;
        return this;
    }

    /**
     * Set a color uniform (RGBA, 0.0-1.0).
     * 设置颜色uniform（RGBA，0.0-1.0）。
     */
    setColor(name: string, r: number, g: number, b: number, a: number = 1.0): this {
        this.uniforms.set(name, { type: UniformType.Color, value: [r, g, b, a] });
        this._dirty = true;
        return this;
    }

    /**
     * Set an integer uniform.
     * 设置整数uniform。
     */
    setInt(name: string, value: number): this {
        this.uniforms.set(name, { type: UniformType.Int, value: Math.floor(value) });
        this._dirty = true;
        return this;
    }

    /**
     * Set a texture uniform.
     * 设置纹理uniform。
     */
    setTexture(name: string, texturePath: string): this {
        this.textures.set(name, texturePath);
        this._dirty = true;
        return this;
    }

    // ============= Uniform Getters =============
    // ============= Uniform获取方法 =============

    /**
     * Get a uniform value.
     * 获取uniform值。
     */
    getUniform(name: string): UniformValue | undefined {
        return this.uniforms.get(name);
    }

    /**
     * Get a texture path.
     * 获取纹理路径。
     */
    getTexture(name: string): string | undefined {
        return this.textures.get(name);
    }

    /**
     * Get all uniforms.
     * 获取所有uniform。
     */
    getUniforms(): Map<string, UniformValue> {
        return this.uniforms;
    }

    /**
     * Get all textures.
     * 获取所有纹理。
     */
    getTextures(): Map<string, string> {
        return this.textures;
    }

    // ============= Serialization =============
    // ============= 序列化 =============

    /**
     * Export to material definition.
     * 导出为材质定义。
     */
    toDefinition(): MaterialDefinition {
        const uniformsObj: Record<string, UniformValue> = {};
        for (const [key, value] of this.uniforms) {
            uniformsObj[key] = value;
        }

        const texturesObj: Record<string, string> = {};
        for (const [key, value] of this.textures) {
            texturesObj[key] = value;
        }

        return {
            name: this.name,
            shader: this.shaderId,
            blendMode: this.blendMode,
            cullMode: this.cullMode,
            depthTest: this.depthTest,
            depthWrite: this.depthWrite,
            uniforms: Object.keys(uniformsObj).length > 0 ? uniformsObj : undefined,
            textures: Object.keys(texturesObj).length > 0 ? texturesObj : undefined
        };
    }

    /**
     * Import from material definition.
     * 从材质定义导入。
     */
    static fromDefinition(def: MaterialDefinition): Material {
        const material = new Material(def.name, typeof def.shader === 'number' ? def.shader : 0);
        material.blendMode = def.blendMode ?? BlendMode.Alpha;
        material.cullMode = def.cullMode ?? CullMode.None;
        material.depthTest = def.depthTest ?? false;
        material.depthWrite = def.depthWrite ?? false;

        if (def.uniforms) {
            for (const [key, value] of Object.entries(def.uniforms)) {
                material.uniforms.set(key, value);
            }
        }

        if (def.textures) {
            for (const [key, value] of Object.entries(def.textures)) {
                material.textures.set(key, value);
            }
        }

        return material;
    }

    // ============= Static Factory Methods =============
    // ============= 静态工厂方法 =============

    /**
     * Create a default sprite material.
     * 创建默认精灵材质。
     */
    static sprite(): Material {
        return new Material('Sprite', 0);
    }

    /**
     * Create an additive (glow) material.
     * 创建加法（发光）材质。
     */
    static additive(): Material {
        const mat = new Material('Additive', 0);
        mat.blendMode = BlendMode.Additive;
        return mat;
    }

    /**
     * Create a multiply (shadow) material.
     * 创建乘法（阴影）材质。
     */
    static multiply(): Material {
        const mat = new Material('Multiply', 0);
        mat.blendMode = BlendMode.Multiply;
        return mat;
    }

    /**
     * Create an unlit/opaque material.
     * 创建无光照/不透明材质。
     */
    static unlit(): Material {
        const mat = new Material('Unlit', 0);
        mat.blendMode = BlendMode.None;
        return mat;
    }
}

File: packages/engine/src/renderer/material/material.rs
//! Material definition and properties.
//! 材质定义和属性。

use super::uniform::MaterialUniforms;

/// Blend modes for material rendering.
/// 材质渲染的混合模式。
#[derive(Clone, Copy, Debug, PartialEq, Eq, Default)]
pub enum BlendMode {
    /// No blending, fully opaque | 无混合，完全不透明
    None,
    /// Standard alpha blending | 标准透明度混合
    #[default]
    Alpha,
    /// Additive blending (good for glow effects) | 加法混合（适用于发光效果）
    Additive,
    /// Multiplicative blending (good for shadows) | 乘法混合（适用于阴影）
    Multiply,
    /// Screen blending (opposite of multiply) | 滤色混合（与乘法相反）
    Screen,
    /// Premultiplied alpha | 预乘透明度
    PremultipliedAlpha,
}

/// Cull modes for material rendering.
/// 材质渲染的剔除模式。
#[derive(Clone, Copy, Debug, PartialEq, Eq, Default)]
pub enum CullMode {
    /// No face culling | 不剔除
    #[default]
    None,
    /// Cull front faces | 剔除正面
    Front,
    /// Cull back faces | 剔除背面
    Back,
}

/// Material definition for 2D rendering.
/// 2D渲染的材质定义。
///
/// A material combines a shader program with uniform parameters and render states.
/// 材质将着色器程序与uniform参数和渲染状态组合在一起。
#[derive(Clone, Debug)]
pub struct Material {
    /// Shader program ID | 着色器程序ID
    pub shader_id: u32,

    /// Material uniform parameters | 材质uniform参数
    pub uniforms: MaterialUniforms,

    /// Blend mode | 混合模式
    pub blend_mode: BlendMode,

    /// Cull mode | 剔除模式
    pub cull_mode: CullMode,

    /// Depth test enabled | 是否启用深度测试
    pub depth_test: bool,

    /// Depth write enabled | 是否启用深度写入
    pub depth_write: bool,

    /// Material name (for debugging) | 材质名称（用于调试）
    pub name: String,
}

impl Default for Material {
    fn default() -> Self {
        Self {
            shader_id: 0, // Default sprite shader
            uniforms: MaterialUniforms::new(),
            blend_mode: BlendMode::Alpha,
            cull_mode: CullMode::None,
            depth_test: false,
            depth_write: false,
            name: "Default".to_string(),
        }
    }
}

impl Material {
    /// Create a new material with default settings.
    /// 使用默认设置创建新材质。
    pub fn new(name: &str) -> Self {
        Self {
            name: name.to_string(),
            ..Default::default()
        }
    }

    /// Create a material with a specific shader.
    /// 使用特定着色器创建材质。
    pub fn with_shader(name: &str, shader_id: u32) -> Self {
        Self {
            name: name.to_string(),
            shader_id,
            ..Default::default()
        }
    }

    /// Set the blend mode.
    /// 设置混合模式。
    pub fn set_blend_mode(&mut self, mode: BlendMode) -> &mut Self {
        self.blend_mode = mode;
        self
    }

    /// Set a float uniform.
    /// 设置浮点uniform。
    pub fn set_float(&mut self, name: &str, value: f32) -> &mut Self {
        self.uniforms.set_float(name, value);
        self
    }

    /// Set a vec2 uniform.
    /// 设置vec2 uniform。
    pub fn set_vec2(&mut self, name: &str, x: f32, y: f32) -> &mut Self {
        self.uniforms.set_vec2(name, x, y);
        self
    }

    /// Set a vec3 uniform.
    /// 设置vec3 uniform。
    pub fn set_vec3(&mut self, name: &str, x: f32, y: f32, z: f32) -> &mut Self {
        self.uniforms.set_vec3(name, x, y, z);
        self
    }

    /// Set a vec4 uniform.
    /// 设置vec4 uniform。
    pub fn set_vec4(&mut self, name: &str, x: f32, y: f32, z: f32, w: f32) -> &mut Self {
        self.uniforms.set_vec4(name, x, y, z, w);
        self
    }

    /// Set a color uniform (RGBA, 0.0-1.0).
    /// 设置颜色uniform（RGBA，0.0-1.0）。
    pub fn set_color(&mut self, name: &str, r: f32, g: f32, b: f32, a: f32) -> &mut Self {
        self.uniforms.set_color(name, r, g, b, a);
        self
    }
}

// ============= Built-in material presets =============
// ============= 内置材质预设 =============

impl Material {
    /// Create a standard sprite material.
    /// 创建标准精灵材质。
    pub fn sprite() -> Self {
        Self::new("Sprite")
    }

    /// Create an additive (glow) material.
    /// 创建加法（发光）材质。
    pub fn additive() -> Self {
        let mut mat = Self::new("Additive");
        mat.blend_mode = BlendMode::Additive;
        mat
    }

    /// Create a multiply (shadow) material.
    /// 创建乘法（阴影）材质。
    pub fn multiply() -> Self {
        let mut mat = Self::new("Multiply");
        mat.blend_mode = BlendMode::Multiply;
        mat
    }

    /// Create an unlit/opaque material.
    /// 创建无光照/不透明材质。
    pub fn unlit() -> Self {
        let mut mat = Self::new("Unlit");
        mat.blend_mode = BlendMode::None;
        mat
    }
}
